### PR TITLE
build: add --no-use-cmake option to configure.py

### PR DESCRIPTION
**Pull request by tchaikov**
as part of the efforts to address scylladb/scylladb#2717, we are
switching over to the CMake-based building system, and fade out the
mechinary to create the rules manually in `configure.py`.

in this change, we add `--no-use-cmake` to `configure.py`, it serves
two purposes:

* prepare for the change which enables cmake by default, by then,
  we would set the default value of `use_cmake` to True, and allow
  user to keep using the existing mechinary in the transition period
  using `--no-use-cmake`.
* allows the CI to tell if a tree is able to build with CMake.
  the command line option of `--use-cmake` is also used by the CI
  workflows, and is passed to `configure.py` if `BUILD_WITH_CMAKE`
  jenkins pipeline parameter is set. but not all branches with
  `--use-cmake` are ready to build with CMake -- only the latest
  master HEAD is ready. so the CI needs to check the capability of
  building with CMake by looking at the output of `configure.py --help`,
  to see if it includes `--no-use-cmake`.
  after this change lands. we will remove the `BUILD_WITH_CMAKE`
  parameter, and use cmake as long as `configure.py` supports
  `--no-use-cmake` option.

the existing mechinary will stay with us for a short transition
period so that developers can take time to get used to the
usage of the naming of targets and the new  directory arrangement.

as a side effect, #20079 will be fixed after switching to CMake.

---

this is a cmake-related change, hence no need to backport.

### CODE REVIEW

File: configure.py
Comment (regevran) [inline configure.py:759]: Can you please clarify how does `--no-use-cmake` does not use cmake when `--use-cmake` is defaulted to `false`?

Comment (tchaikov) [reply configure.py:759]: @regevran did you get a chance to read the commit message? i am happy to explain it in more details if you have a minute.

**Review by regevran — APPROVED** [review]
I think you should squash the two commits together.

**Comment by tchaikov** [comment]
> I think you should squash the two commits together.

why? could you help me understand why you prefer squashing them?

File: configure.py
Comment (tchaikov) [reply configure.py:767]: yes, but the usage i want to address is like:
```console
$ ./configure.py --foobar --help
usage: Configure scylla [-h] [--out BUILDFILE] [--out-final-name BUILDFILE_FINAL_NAME] [--mode {debug,release,dev,sanitize,coverage}] [--with ARTIFACTS] [--with-seastar SEASTAR_PATH] [--enable-dist]
                        [--disable-dist] [--dist-only] [--cflags USER_CFLAGS] [--ldflags USER_LDFLAGS] [--target TARGET] [--compiler CXX] [--c-compiler CC] [--enable-dpdk] [--disable-dpdk]
                        [--dpdk-target DPDK_TARGET] [--debuginfo DEBUGINFO] [--optimization-level MODE=LEVEL] [--static-stdc++] [--static-boost] [--static-yaml-cpp] [--tests-debuginfo TESTS_DEBUGINFO]
                        [--perf-tests-debuginfo PERF_TESTS_DEBUGINFO] [--python PYTHON] [--split-dwarf] [--enable-alloc-failure-injector] [--enable-seastar-debug-allocations] [--with-antlr3 ANTLR3_EXEC]
                        [--with-ragel RAGEL_EXEC] [--enable-stack-guards] [--disable-stack-guards] [--verbose] [--test-repeat TEST_REPEAT] [--test-timeout TEST_TIMEOUT]
                        [--clang-inline-threshold CLANG_INLINE_THRESHOLD] [--list-artifacts] [--date-stamp DATE_STAMP] [--use-cmake] [--coverage] [--build-dir BUILD_DIR]
...
$ echo $?
0
```

`--help` is added so that the given call is always a noop.

`argparse` invokes the action registered by the added options, if the action calls `sys.exit()`. the code which checks for unknown args is not executed. briefly, it works like:
```py
    def parse_args(self, args=None, namespace=None):
        # set the value to the namespace to be returned, 
        # calls into the registered actions, and
        # consume the positional options, 
        args, argv = self.parse_known_args(args, namespace)
        if argv:
            msg = _('unrecognized arguments: %s') % ' '.join(argv)
            if self.exit_on_error:
                self.error(msg)
            else:
                raise ArgumentError(None, msg)
        return args
```

**Comment by regevran** [comment]
> > I think you should squash the two commits together.
> 
> why? could you help me understand why you prefer squashing them?

Just because they both change the same file. And yes, logically one can argue that the two changes are unrelated, so no need to squash.
Anyway, it doesn't matter, you don't have to, both ways are fine.

**Comment by tchaikov** [comment]
> > > I think you should squash the two commits together.
> > 
> > 
> > why? could you help me understand why you prefer squashing them?
> 
> Just because they both change the same file. And yes, logically one can argue that the two changes are unrelated, so no need to squash. Anyway, it doesn't matter, you don't have to, both ways are fine.

i disagree with you. they do change the same file. but this does not imply they serve the same purpose. please consider, if we have a project which is small enough, and it only have 3 source files, does this make us merge all PRs into 3 ?

**Comment by denesb** [comment]
>  to see if it includes --no-use-cmake. after this change lands. we will remove the BUILD_WITH_CMAKEparameter, and use cmake as long asconfigure.pysupports--no-use-cmake` option.

backticks are off in the cover letter.

**Comment by tchaikov** [comment]
> > to see if it includes --no-use-cmake. after this change lands. we will remove the BUILD_WITH_CMAKEparameter, and use cmake as long asconfigure.pysupports--no-use-cmake` option.
> 
> backticks are off in the cover letter.

@denesb thanks for pointing this out. fixed.

**Comment by tchaikov** [comment]
i'd like to have this change in master after #20288 lands.

**Comment by tchaikov** [comment]
v2:

- resolve the conflicts with master HEAD

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help review this change?

**Review by denesb — APPROVED** [review]

**Pull request merged**
